src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -165,18 +182,39 @@ public override void AdvanceTo(SequencePosition consumed)
 
         public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
         {
-            if (_inputLength == 0)
+            if (!_isReading)
             {
-                return;
+                throw new InvalidOperationException("No reading operation to complete.");
             }
 
-            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+            _isReading = false;
+
+            if (_readCompleted)
+            {
+                _readResult = new ReadResult(_readResult.Buffer.Slice(consumed, _readResult.Buffer.End), isCanceled: false, _readCompleted);
 
-            _inputLength -= dataLength;
+                if (!_finalAdvanceAfterReadingEntireContentLength)
+                {
+                    _context.Input.AdvanceTo(consumed, examined);
```

Comment:
What if the app didn't consume/examine everything the first time it gets a completed ReadResult? Wouldn't you end up never advancing the connection pipe to the end of the request body? I think this should be the following:

```suggestion
                    _context.Input.AdvanceTo(_readResult.Buffer.End);
```

Reply:
Don't actually make the above change, because the app code will likely still use the buffers. Now I think we should always call `_context.Input.AdvanceTo(consumed, examined);` and just get rid of _finalAdvanceAfterReadingEntireContentLength.

Reply:
We can't call AdvanceTo more than once because we stop calling ReadAsync on the pipe once we reach the content length.